### ISSUE TITLE
ChebyKAN Having troubles on solving dynamical systems

### ISSUE THREAD

**Issue by mccrow2018** [SynodicMonth/ChebyKAN#4]
Hi, ChebyKAN is indeed simple, elegant and powerful, I believe it can do more.

So I implemented it on solving some models with dynamical systems, economic models to be precise, where the dynamical systems or equations quite similar to PDEs.

The main problem I encountered is that ChebyKAN is more prone to be "stuck", preventing training going any further. Here are two illustrations with KAN structure

```
valuefunction_KAN  =  KAN(width=[2,5,5,1], grid=5, k=3, grid_eps=1.0, noise_scale_base=0.25)
```

and ChebyKAN structure

```
class ChebyKAN(nn.Module):
    def __init__(self):
        super(ChebyKAN, self).__init__()
        self.chebykan1 = ChebyKANLayer(2, 8, 8)
        self.chebykan2 = ChebyKANLayer(8, 16, 5)
        self.chebykan3 = ChebyKANLayer(16, 1, 5)

    def forward(self, x):
        x = self.chebykan1(x)
        x = self.chebykan2(x)
        x = self.chebykan3(x)
        return x

valuefunction_cheb  =  ChebyKAN()
```

Results on the first fig are trained by LBFGS, and by Adam with learning rate 1e-2 on the second fig.

![aa3dee3b0229880fe6b2e1e9b2daa04](https://github.com/SynodicMonth/ChebyKAN/assets/35625194/750dd1ec-06da-4859-ac01-bcf9309753ed)
![5556b5a73690251bc5af3c1a4524213](https://github.com/SynodicMonth/ChebyKAN/assets/35625194/a8bacc10-3c55-4615-8925-579af85e0aff)

I have tested it multiple times, with different input, output dimensions and degree range from 4 to 12, the issue remains.

**Comment by JanRocketMan** [comment]
I believe that's expected due to the usage of tanh nonlinearity since Chebyshev polynomials are defined on [-1, 1] range.

Have you tried adding the residuals to the layers, or that's smth you wish to avoid? E.g. this way:

```python
class ResidualChebyKAN(nn.Module):
    def __init__(self):
        super().__init__()
        self.chebykan1 = ChebyKANLayer(2, 8, 8)
        self.chebykan2 = ChebyKANLayer(8, 16, 5)
        self.chebykan3 = ChebyKANLayer(16, 1, 5)

    def forward(self, x):
        x = self.chebykan1(x) + x
        x = self.chebykan2(x) + x
        x = self.chebykan3(x) + x
        return x

valuefunction_cheb  =  ResidualChebyKAN()
```

**Comment by mccrow2018** [comment]
Thanks for the advice @JanRocketMan. I have tried this method and implemented it like the structure in paper, residuals in SiLU function plus approximations, I even added learnable scales. This did not solve the issue, but even be worse.

modified ChebyKAN:

```
import torch
import torch.nn as nn
import numpy as np

# This is inspired by Kolmogorov-Arnold Networks but using Chebyshev polynomials instead of splines coefficients
class ChebyKANLayer(nn.Module):
    def __init__(self, input_dim, output_dim, degree):
        super(ChebyKANLayer, self).__init__()
        self.inputdim      =  input_dim
        self.outdim        =  output_dim
        self.degree        =  degree

        self.cheby_coeffs  =  nn.Parameter(torch.empty(input_dim, output_dim, degree + 1))
        self.scale_base    =  nn.Parameter(torch.ones(input_dim * output_dim))
        self.scale_cheb    =  nn.Parameter(torch.ones(input_dim * output_dim))

        nn.init.normal_(self.cheby_coeffs, mean=0.0, std=1/(input_dim * (degree + 1)))

    def forward(self, x):
        x  =  torch.reshape(x, (-1, self.inputdim))  # shape = (batch_size, inputdim)

        xb = torch.einsum('bi,o->boi', x, torch.ones(self.outdim)).reshape(-1, self.inputdim * self.outdim)
        xb = (xb / (1 + torch.exp(-xb))) * self.scale_base.unsqueeze(dim = 0)
        xb = torch.sum(xb.reshape(x.shape[0], self.outdim, self.inputdim), dim = 2)

        x  =  torch.tanh(x) 
        
        # Initialize Chebyshev polynomial tensors
        cheby  =  torch.ones(x.shape[0], self.inputdim, self.degree + 1, device = x.device)

        if self.degree > 0:
            cheby[:, :, 1]  =  x
        for i in range(2, self.degree + 1):
            cheby[:, :, i]  =  2 * x * cheby[:, :, i - 1].clone() - cheby[:, :, i - 2].clone()
        # Compute the Chebyshev interpolation
        y  =  torch.einsum('bid,iod->bio', cheby, self.cheby_coeffs).reshape(-1, self.inputdim * self.outdim)
        y  =  y * self.scale_cheb.unsqueeze(dim = 0)
        y  =  torch.sum(y.reshape(x.shape[0], self.inputdim, self.outdim), dim = 1) 
        y  =  y.view(-1, self.outdim) + xb

        return y
```

and the results trained by LBFGS:

![image](https://github.com/SynodicMonth/ChebyKAN/assets/35625194/5c9b1143-6813-4bfd-962e-70908e57fe1e)

**Comment by SynodicMonth** [comment]
Its probably caused by tanh or/and the [-1, 1] domin. is it possible to provide the equation of the system or any code snippets? im excited to test on it.

**Comment by mccrow2018** [comment]
It would be an honor. Here is a Colab link of the snippet: [KAN-sto-test].

**Comment by SynodicMonth** [comment]
I mede several tests and tunes and it seems ChebyKAN is truly a disaster on this problem. ChebyKAN [1,5,5,1] @ degree of 5 with @JanRocketMan 's advice indeed improves the performace in some way but its nowhere near the original KAN.
I assume there're some fundamental flaws in the cheby. I think i need to test other polys and find a way to replace tanh.
My apologies for the buggy implemention. ill be back when i found the solution.

**Comment by Boris-73-TA** [comment]
I used this alternative: 
class OrthogPolyKAN(nn.Module):
    def __init__(self):
        super(OrthogPolyKAN, self).__init__()
        self.orthogpolykan1 = OrthogPolyKANLayer(2, 5, 5, poly_type='Jacobi', alpha=3, beta=3)
        self.orthogpolykan2 = OrthogPolyKANLayer(5, 8, 4, poly_type='Cheby2')
        self.orthogpolykan3 = OrthogPolyKANLayer(8, 1, 3) #  Legendre is default poly_type

    def forward(self, x):
        x = self.orthogpolykan1(x)
        x = self.orthogpolykan2(x)
        x = self.orthogpolykan3(x)
        return x

valuefunction_op  =  OrthogPolyKAN() 
I get the following plot: 
<img width="473" alt="Screenshot 2024-05-14 at 21 07 25" src="https://github.com/SynodicMonth/ChebyKAN/assets/129144076/79f079cd-40d2-4230-a7b7-a5e9b864f711">
Maybe including layers of different polynomials helps?

**Comment by SynodicMonth** [comment]
> I used this alternative: class OrthogPolyKAN(nn.Module): def **init**(self): super(OrthogPolyKAN, self).**init**() self.orthogpolykan1 = OrthogPolyKANLayer(2, 5, 5, poly_type='Jacobi', alpha=3, beta=3) self.orthogpolykan2 = OrthogPolyKANLayer(5, 8, 4, poly_type='Cheby2') self.orthogpolykan3 = OrthogPolyKANLayer(8, 1, 3) # Legendre is default poly_type
> 
> ```
> def forward(self, x):
>     x = self.orthogpolykan1(x)
>     x = self.orthogpolykan2(x)
>     x = self.orthogpolykan3(x)
>     return x
> ```
> 
> valuefunction_op = OrthogPolyKAN() I get the following plot: <img alt="Screenshot 2024-05-14 at 21 07 25" width="473" src="https://private-user-images.githubusercontent.com/129144076/330577986-79f079cd-40d2-4230-a7b7-a5e9b864f711.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTU4NjQwNjgsIm5iZiI6MTcxNTg2Mzc2OCwicGF0aCI6Ii8xMjkxNDQwNzYvMzMwNTc3OTg2LTc5ZjA3OWNkLTQwZDItNDIzMC1hN2I3LWE1ZTliODY0ZjcxMS5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwNTE2JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDUxNlQxMjQ5MjhaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02M2Q4OWY3MWZhODFiYzZjNTRhZjdiZjRiNzAwM2UzZmI3MDlkMjViODk5MTI1MDY5M2QwN2U1MDk0NmFhYjY5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.MVlKQhH8m3Zhcoyh-ptHh0rvrHgfZco_rmSf_dKsVDw"> Maybe including layers of different polynomials helps?

That's a significant improvement. I indeed found that Chebyshev polynomials have limitations in the properties of the functions they can model. Additionally, I have another idea: it might be because ChebyKAN does not support continual learning (see #8). I'm not very familiar with these economic models, so I'm not sure if the input range will change if the model's prediction changes (I see many iterations from the same value function). 
I've been sick recently. I'll run more tests once I recover.
